app/scripts/offscreen.js

```diff
@@ -0,0 +1,42 @@
+import { OffscreenCommunicationTarget } from '../../shared/constants/offscreen-communication';
+
+/**
+ * Creates an offscreen document that can be used to load additional scripts
+ * and iframes that can communicate with the extension through the chrome
+ * runtime API. Only one offscreen document may exist, so any iframes required
+ * by extension can be embedded in the offscreen.html file. See the offscreen
+ * folder for more details.
+ */
+export async function createOffscreen() {
+  const { chrome } = globalThis;
+  if (!chrome.offscreen || (await chrome.offscreen.hasDocument())) {
+    return;
+  }
+
+  const loadPromise = new Promise((resolve) => {
+    chrome.runtime.onMessage.addListener((msg) => {
+      if (
+        msg.target === OffscreenCommunicationTarget.extensionMain &&
+        msg.isBooted
+      ) {
+        resolve();
+      }
+    });
+  });
+
+  await chrome.offscreen.createDocument({
+    url: './offscreen.html',
+    reasons: ['IFRAME_SCRIPTING'],
+    justification:
+      'Used for Hardware Wallet and Snaps scripts to communicate with the extension.',
+  });
+
+  // In case we are in a bad state where the offscreen document is not loading, timeout and let execution continue.
+  const timeoutPromise = new Promise((resolve) => {
+    setTimeout(resolve, 5000);
```

Comment:
I somehow have a feeling that we might want to increase this timeout tolerance 🤔

On a slower machine (in some situations), chances are that this timeout can be reached and offscreen not loaded. In that case, it would not be loaded just because of slowness, not because of functional failure, and the slowness factor should be tolerated little bit more IMO.

Maybe 7-10 seconds would be better? WDYT?

Reply:
I'm expecting this load time to be fast since it is strictly waiting for scripts to be loaded from disk and executed. Essentially the promise should resolve as soon as LavaMoat has run and the offscreen.ts file has fully executed. We are not waiting for the network here.

Reply:
Yea, fair enough. I know this is not related to network, but still, my concern was if a machine running this is slower and under heavy pressure (memory, cpu, etc.), it might not be capable of loading this just in time required and get the message, especially because the offscreen is some type of detached process. It would be an edge case anyway.

It's alright if you're confident about the execution time and max timeout of 5 seconds.